nupic/research/frameworks/pytorch/model_utils.py

```diff
@@ -111,8 +118,25 @@ def train_model(
             pre_batch_callback(model=model, batch_idx=batch_idx)
 
         optimizer.zero_grad()
+
         output = model(data)
-        loss = criterion(output, target)
+
+        # alternative knowledge distillation training
+        if teacher_model is not None:
+            with torch.no_grad():
+                # target is linear combination of teacher and target softmaxes
+                softmax_output_teacher = F.softmax(teacher_model(data))
+                one_hot_target = F.one_hot(target, num_classes=output.shape[-1])
+                combined_target = (kd_factor * softmax_output_teacher
+                                   + (1 - kd_factor) * one_hot_target)
+            # requires a custom loss function.
+            del softmax_output_teacher, one_hot_target
+            loss = soft_cross_entropy(output, combined_target)
+            del combined_target
+        # regular training
```

Comment:
This code seems to be very specific for this specific algorithm to be part of this generic training function. 
ImagenetExperiment has an option for custom trainers ([train_model_func](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py#L157-L159)) that can be used for special training cases like this one. 
if you like I can point you to various experiments Michaelangelo created using custom trainers that you can use a reference.